analysis/src/ip_protocol/ip_protocol.rs

```diff
@@ -12,3 +12,176 @@ pub trait Packet {
+        }
+
+        fn as_any(&self) -> &dyn std::any::Any {
+            self
+        }
+    }
+
+    pub fn new_packet(data: &[u8]) -> PcapPacket<'_> {
+        PcapPacket {
+            header: &PacketHeader {
+                ts: libc::timeval {
+                    tv_sec: 0 as libc::time_t,
+                    tv_usec: 0 as libc::time_t,
+                },
+                caplen: 0,
+                len: 0,
+            },
+            data,
+        }
+    }
+
+    #[test]
+    fn packet_filter_remove_packet() {
+        let mut data = [0u8; 60];
+        data[14] = 0x5;
+
+        let packet = new_packet(&data);
+        let eth_frame = EthernetFrame::serialize(&packet);
+        let mut ip_packet = Some(IPv4::serialize(&packet, eth_frame));
+
+        ip_packet = packet_filter::<IPv4>(ip_packet, |p| p.ihl == 20);
+
+        if let Some(packet) = ip_packet {
+            let packet_ref: &dyn Packet = &*packet;
+            let ipv4_packet: &IPv4 = packet_ref
+                .as_any()
+                .downcast_ref::<IPv4>()
+                .expect("Not IPv4");
+
+            assert_eq!(ipv4_packet.ihl, 20);
+        } else {
+            assert!(false);
+        }
+    }
+
+    #[test]
+    fn packet_filter_keep_packet() {
+        let mut data = [0u8; 60];
+        data[14] = 0x5;
+
+        let packet = new_packet(&data);
+        let eth_frame = EthernetFrame::serialize(&packet);
+        let mut ip_packet = Some(IPv4::serialize(&packet, eth_frame));
+
+        ip_packet = packet_filter::<IPv4>(ip_packet, |_p| false);
+
+        if let Some(_packet) = ip_packet {
+            assert!(false);
+        } else {
+            assert!(true);
+        }
+    }
+
+    #[test]
+    fn packet_filter_keep_packet_when_filter_other_protocol() {
+        let mut data = [0u8; 60];
+        data[14] = 0x5;
+
+        let packet = new_packet(&data);
+        let eth_frame = EthernetFrame::serialize(&packet);
+        let mut ip_packet = Some(IPv4::serialize(&packet, eth_frame));
+
+        ip_packet = packet_filter::<OtherProtocol>(ip_packet, |_p| true);
+        ip_packet = packet_filter::<IPv4>(ip_packet, |p| p.ihl == 20);
+        ip_packet = packet_filter::<OtherProtocol>(ip_packet, |_p| true);
+
+        if let Some(packet) = ip_packet {
+            let packet_ref: &dyn Packet = &*packet;
+            let ipv4_packet: &IPv4 = packet_ref
+                .as_any()
+                .downcast_ref::<IPv4>()
+                .expect("Not IPv4");
+
+            assert_eq!(ipv4_packet.ihl, 20);
+        } else {
+            assert!(false);
+        }
+    }
+
+    #[test]
+    fn packet_filter_remove_packet_when_filter_other_protocol() {
+        let mut data = [0u8; 60];
+        data[14] = 0x5;
+
+        let packet = new_packet(&data);
+        let eth_frame = EthernetFrame::serialize(&packet);
+        let mut ip_packet = Some(IPv4::serialize(&packet, eth_frame));
+
+        ip_packet = packet_filter::<OtherProtocol>(ip_packet, |_p| true);
```

Comment:
Hvorfor bliver det gjort to gange for OtherProtocol?

Reply:
Same